pkg/resource/deploy/plan_executor.go

```diff
@@ -0,0 +1,247 @@
+		return ErrCanceled
+	}
+
+	if pe.errored() {
+		logging.V(planExecutorLogLevel).Infof("PlanExecutor.Execute(...): observed that the plan errored")
+		if pe.preview {
+			return ErrPreviewFailed
+		}
+
+		return ErrUpdateFailed
+	}
+
+	logging.V(planExecutorLogLevel).Infof("PlanExecutor.Execute(...): observed that the plan was successful")
+	return nil
+}
+
+// Summary returns a PlanSummary of the plan that was executed.
+func (pe *PlanExecutor) Summary() PlanSummary {
+	return pe.stepGen
+}
+
+// errored returns whether or not this plan failed due to an error in step application.
+func (pe *PlanExecutor) errored() bool {
+	return pe.sawError.Load().(bool) || pe.stepExec.Errored()
+}
+
+// canceled returns whether or not this plan failed because it was canceled through Ctrl-C.
+func (pe *PlanExecutor) canceled() bool {
+	return pe.sawCancel.Load().(bool)
+}
+
+// cancelDueToError cancels the step executor and signals shutdown because the plan executor witnessed
+// an error that the step executor would not have witnessed. The main reason this happens is because of errors
+// occurring in the source program that can't be translated into chains for the step executor to execute.
+func (pe *PlanExecutor) cancelDueToError() {
+	pe.sawError.Store(true)
+	pe.cancel()
+}
+
+// handleSingleEvent handles a single source event. For all incoming events, it produces a chain that needs
+// to be executed and schedules the chain for execution.
+func (pe *PlanExecutor) handleSingleEvent(event SourceEvent) {
+	if event == nil {
+		logging.V(planExecutorLogLevel).Infof("PlanExecutor.handleSingleEvent(...): received nil event")
+		return
+	}
+
+	logging.V(planExecutorLogLevel).Infof("PlanExecutor.handleSingleEvent(...): received event")
+	switch e := event.(type) {
+	case RegisterResourceEvent:
+		step, steperr := pe.stepGen.GenerateSteps(e)
+		if steperr != nil {
+			logging.V(planExecutorLogLevel).Infof(
+				"PlanExecutor.handleSingleEvent(...): received step event error: %v", steperr.Error())
+			pe.cancel()
+			return
+		}
+
+		logging.V(planExecutorLogLevel).Infof("PlanExecutor.handleSingleEvent(...): submitting chain for execution")
+		pe.stepExec.Execute(step)
+	case ReadResourceEvent:
+		step, steperr := pe.stepGen.GenerateReadSteps(e)
+		if steperr != nil {
+			logging.V(planExecutorLogLevel).Infof(
+				"PlanExecutor.handleSingleEvent(...): received step event error: %v", steperr.Error())
+			pe.cancel()
+			return
+		}
+
+		logging.V(planExecutorLogLevel).Infof("PlanExecutor.handleSingleEvent(...): submitting reads for execution")
+		pe.stepExec.Execute(step)
+	case RegisterResourceOutputsEvent:
+		logging.V(planExecutorLogLevel).Infof("PlanExecutor.handleSingleEvent(...): received register resource outputs")
+		pe.stepExec.ExecuteRegisterResourceOutputs(e)
+	}
+}
+
+// NewPlanExecutor creates a new PlanExecutor suitable for executing the given plan.
+func NewPlanExecutor(parentCtx context.Context, plan *Plan, opts Options,
```

Comment:
Non-blocking: rather than:
```
p := NewPlan(...)
e := NewPlanExecutor(ctx, plan, ...)
err := e.Execute()
```

I'd much rather see
```
p := NewPlan(...)
err := p.Execute()
```

I can't think of a strong argument for exposing the `PlanExecutor` as its own type (likewise the `StepExecutor`).

Reply:
I just tried to do this real quick and it's a little invasive, I'll do this as a follow-up with https://github.com/pulumi/pulumi/issues/1712 since killing off `plan.Walk` will help with error reporting.